src/Gaffer/Reference.cpp

```diff
@@ -137,15 +138,139 @@ void transferOutputs( Gaffer::Plug *srcPlug, Gaffer::Plug *dstPlug )
 
 } // namespace
 
+//////////////////////////////////////////////////////////////////////////
+// MetadataEdits. This internal utility class is used to track where edits have
+// been applied to plugs following loading.
+//////////////////////////////////////////////////////////////////////////
+
+class Reference::MetadataEdits : public boost::signals::trackable
+{
+
+	public :
+
+		MetadataEdits( Reference *reference )
+			:	m_reference( reference )
+		{
+		}
+
+		bool hasMetadataEdit( const Plug *plug, const InternedString &key ) const
+		{
+			const MetadataEdit *edit = metadataEdit( plug, key );
+			return edit ? edit->metadataSet : false;
+		}
+
+		void connect()
+		{
+			m_connection = Metadata::plugValueChangedSignal().connect( boost::bind( &MetadataEdits::plugValueChanged, this, ::_1, ::_2, ::_3, ::_4 ) );
+		}
+
+		void disconnect()
+		{
+			m_connection.disconnect();
+		}
+
+	private :
+
+		Reference *m_reference;
+		boost::signals::scoped_connection m_connection;
+
+		struct MetadataEdit
+		{
+			MetadataEdit()
+				:	metadataSet( false )
+			{
+			}
+			bool metadataSet;
+		};
+
+		using MetadataEditKey = std::pair<const std::string, const InternedString>;
+		using MetadataEditsMap = std::map<MetadataEditKey, MetadataEdit>;
+		MetadataEditsMap m_metadataEdits;
+
+		const MetadataEdit *metadataEdit( const Plug *plug, const InternedString &key ) const
+		{
+			// Cheeky cast better than maintaining two near-identical functions.
+			return const_cast<MetadataEdits *>( this )->metadataEdit( plug, key, /* createIfMissing = */ false );
+		}
+
+		MetadataEdit *metadataEdit( const Plug *plug, const InternedString &key, bool createIfMissing )
+		{
+			if( plug->ancestor<Reference>() != m_reference )
```

Comment:
af8ee6a

Re `isReferencePlug()`: This is a bit tricky, isn't it? I guess in theory we should still check whether we're messing with the user plug? The problem I found with `isReferencePlug()` is that when we copy the metadata from the old plugs to the new plugs after reloading the grf file, the old plug is prefixed with `__`, suggesting that it's not a reference plug anymore, causing us to ignore the edits. That in turn means we don't have edits on that plug and there's nothing to copy over.

Reply:
Is the situation improved at all now we're indexing by pointer rather than name? Can we put the `isReferencePlug()` check _after_ the lookup for existing edits but _before_ the creation of a new edit? So we still get to query the old edits when we want to transfer them, but we don't create edits for user plugs?

While we're on the subject of the user plug, will we need to do something else in the serialiser to make sure we still serialise metadata on user plugs, even though they're not considered to be edits?